Wikipedia/Code/NotificationsCenterViewController.swift

```diff
@@ -286,31 +325,121 @@ extension NotificationsCenterViewController: UICollectionViewDelegate {
+                break
+            }
+
+            let swipeStackWidth = cell.swipeActionButtonStack.frame.size.width
+            var totalTranslationX = translationX + (cellSwipeData.activelyPannedCellTranslationX ?? 0)
+
+            let maximumTranslationX = swipeStackWidth + swipeEdgeBuffer
+
+            // The user is trying to pan too far left
+            if totalTranslationX < -maximumTranslationX {
+                totalTranslationX = -maximumTranslationX - log(abs(translationX))
+            }
+
+            // Extends too far right
+            if totalTranslationX > swipeEdgeBuffer {
+                totalTranslationX = swipeEdgeBuffer + log(abs(translationX))
+            }
+
+            let finalTranslationX = isRTL ? -totalTranslationX : totalTranslationX
+            cell.foregroundContentContainer.transform = CGAffineTransform(translationX: finalTranslationX, y: 0)
         case .ended:
-            userDidSwipeCell(indexPath: activelyPannedCellIndexPath)
-            activelyPannedCellIndexPath = nil
+            guard let cell = cellSwipeData.activeCell(in: notificationsView.collectionView) else {
+                break
+            }
+
+            var shouldOpenSwipePanel: Bool
+            let currentCellTranslationX = isRTL ? -cell.foregroundContentContainer.transform.tx : cell.foregroundContentContainer.transform.tx
+
+            if currentCellTranslationX > 0 {
+                shouldOpenSwipePanel = false
+            } else {
+                if velocityX < -triggerVelocity {
+                    shouldOpenSwipePanel = true
+                } else {
+                    shouldOpenSwipePanel = abs(currentCellTranslationX) > (0.5 * cell.swipeActionButtonStack.frame.size.width)
+                }
+            }
+
+            if velocityX > triggerVelocity {
+                shouldOpenSwipePanel = false
+            }
+
+            if !shouldOpenSwipePanel {
+                cellSwipeData.resetActiveData()
+            }
+
+            animateSwipePanel(open: shouldOpenSwipePanel, for: cell)
         default:
             return
         }
     }
 
-    /// TODO: This will be removed in the final implementation
-    fileprivate func userDidSwipeCell(indexPath: IndexPath?) {
-        guard let indexPath = indexPath,
-              let cellViewModel = dataSource?.itemIdentifier(for: indexPath) else {
+}
+
+//MARK: NotificationCenterCellDelegate
+
+extension NotificationsCenterViewController: NotificationsCenterCellDelegate {
+
+    func userDidTapSecondaryActionForViewModel(_ cellViewModel: NotificationsCenterCellViewModel) {
+        guard let url = cellViewModel.secondaryURL(for: viewModel.configuration) else {
             return
         }
-        
+        navigate(to: url)
+    }
+
+    func userDidTapMoreActionForCell(_ cell: NotificationsCenterCell) {
+        guard let cellViewModel = cell.viewModel, let indexPath = dataSource?.indexPath(for: cellViewModel) else  {
```

Comment:
Seems like this indexPath is only used to figure out the cell again below in the old popover logic. Can you just reference the cell param in line 477 and remove this?